In [1]:
import numpy as np
import torch

# To create the memory of model we are going to average the past context  
* We are gonna for every single batch element independently and every token in that sequence
* We calculate average of all vectors in all previous tokens and at also current token
* Thats how we get bag of words

# In function we are iterating over all batch-dims independently
* Then iterating over time
* And then the previous tokens are at this batch-dim (x[b, and then everything up to and including t-token :t+1)

In [6]:
torch.manual_seed(42)
a = torch.ones(3,3)
b = torch.randint(0,10,(3,2)).float()
c = a @ b
print('a=')
print(a)
print('b=')
print(b)
print('--')
print('c=')
print(c)

a=
tensor([[1., 1., 1.],
        [1., 1., 1.],
        [1., 1., 1.]])
b=
tensor([[2., 7.],
        [6., 4.],
        [6., 5.]])
--
c=
tensor([[14., 16.],
        [14., 16.],
        [14., 16.]])


# We have simple 3 by 3 matrix a, then matrix b 3 by 2 with random numbers
* And then c matrix, which is matrix multiplication of a and b, which gives us 3 by 2

# So 14 in tensor c is dop product of first row of a [1,1,1] with first column of b [2,6,6]
* And as well 16 in tensor c is dot product between first row of a [1,1,1] with second column of b [7,4,5]

In [4]:
torch.tril(torch.ones(3,3))

tensor([[1., 0., 0.],
        [1., 1., 0.],
        [1., 1., 1.]])

# Then we call torch.trill, which gives us triangular shape tensor

In [7]:
a = torch.tril(torch.ones(3,3))
c = a @ b
print('a=')
print(a)
print('b=')
print(b)
print('--')
print('c=')
print(c)

a=
tensor([[1., 0., 0.],
        [1., 1., 0.],
        [1., 1., 1.]])
b=
tensor([[2., 7.],
        [6., 4.],
        [6., 5.]])
--
c=
tensor([[ 2.,  7.],
        [ 8., 11.],
        [14., 16.]])


# And now when we make multiplication it it gives us following numbers:
* Number 2 in tensor c is dot product between first row of matrix a [1,0,0] and first column of matrix b [2,6,6]
* And next 7 is dot product between first row of matrix a [1,0,0] and second column of matrix b [7,4,5]

# What we get in tensor c first row [2,7], we pluck out first row of b to tensor c
* And next dot product between second row of a and 2 columns of b gives us 2+6 and 7+4, and this is what we want to achieve
* Now depending how many zeros and ones we have in a, we are doing sum of variable number of the rows in b
* And thats is deposited in c

In [8]:
a = a / torch.sum(a, 1, keepdim=True)
c = a @ b
print('a=')
print(a)
print('b=')
print(b)
print('--')
print('c=')
print(c)

a=
tensor([[1.0000, 0.0000, 0.0000],
        [0.5000, 0.5000, 0.0000],
        [0.3333, 0.3333, 0.3333]])
b=
tensor([[2., 7.],
        [6., 4.],
        [6., 5.]])
--
c=
tensor([[2.0000, 7.0000],
        [4.0000, 5.5000],
        [4.6667, 5.3333]])


# Now i we want to take an average of row in b by normalizing rows in a
* We take an a and divide it by sum of a in 1-dimension and keepdims
* Now the rows in a are sum to 1: first row [1.0,0,0] 2-row [0.5, 0.5.0] 3-row [0.3, 0.3, 0.3]
* And when we multiply a and b, we get in c:
* In first row of c [2,7] the first row of b [2,7]
* In second of c [4, 5.5] row we get average of first 2 rows of b [2, 7],[6,4]